### ISSUE TITLE
Build failure in doc/user-manual/language/cubical.lagda

### ISSUE THREAD

**Issue by rwe** [agda/agda#4805]
Current `master` (9c9f0327a44133fd77b1dd4da8b75d4285410e9c) on Travis. For example: https://travis-ci.org/github/agda/agda/jobs/711106308#L1754

The issue was introduced in #4781 (_2ltt: Ported cubical primitives to use strict universes, and other fixes_) and there is some troubleshooting going on there. However to avoid the invisible commentary on closed PRs, I figured it would be worth opening an issue to track it, which can be closed once fixed.

The error is:
```diff
diff --git a/tmp/language-cubical2002-0.golden b/tmp/language-cubical2002-1.actual
index 699465f..28fb2ee 100644
--- a/tmp/language-cubical2002-0.golden
+++ b/tmp/language-cubical2002-1.actual
@@ -1,2 +1,10 @@
-AGDA_SUCCESS
+AGDA_UNEXPECTED_FAIL
 
+ret > ExitFailure 42
+out > agda-default-include-path/Agda/Builtin/Cubical/HCompU.agda:73,11-163
+out > SSetω != SSet l
+out > when checking that the inferred type of an application
+out >   (i : I) → e (~ i)
+out > matches the expected type
+out >   (i : I) → e (~ i)
+out >
```
(The actual message uses the decimal Unicode values `\969` for `ω` and `\8695` for `→`).

This appears to be during type-checking the examples in `doc/user-manual/language/cubical.lagda.rst` and `src/data/lib/prim/Agda/Builtin/Cubical/HCompU.agda` which both changed.

That should be locally testable with `make user-manual-test`, however @asr noted that this did not seem to reproduce the error.

**Comment by rwe** [comment]
I wonder if it potentially could be an issue with either the intermodule propagation or interface-file persistence/compatibility-detection of that new `--two-level` flag? @Saizan @asr does that sound plausible?

If it appears to work locally, that seems to indicate it might not just be a broken example (unless something about the local test skipped a step). Inconsistency between environments/runs makes me suspicious of those interface flags, as I know there have been problems of that kind before, and it's something that changed in that PR.

**Comment by Saizan** [comment]
@rwe the commit also bumps the interface version i.e. `currentInterfaceVersion` so that shouldn't be a problem? Though it might be an idea to bump it again.

**Comment by rwe** [comment]
What I'm wondering is not about compatibility between Agda commits, but between sets of compilation flags. If e.g. `HCompU.agda` generates `HCompU.agdai` with `--two-level` active, does it recognize that that interface file cannot be used when the flag is inactive? (And in the reverse case).

**Comment by Saizan** [comment]
`--two-level` is not infective, maybe it should be, however the modules have `--cubical` on anyway which implies `--two-level`.

The issue might be a different order of unification though, in #4807 I'm checking if there's a more robust formulation of `g`.

**Comment by fredrikNordvallForsberg** [comment]
Actually @Saizan you did make it infective:
https://github.com/agda/agda/blob/b51630a991e3b61e8aafd649f5b96ca23e80958b/src/full/Agda/Interaction/Options.hs#L486-L489
:slightly_smiling_face:

**Comment by rwe** [comment]
If #4807 (_eta expand use of `fill`_) fixes the build, I worry that might be covering up a potential real issue. (Unless there's already an underlying issue about dependence of type-checking order that wasn't caused by this. Is there? Maybe @andreasabel or @nad would know?)

Seeing the code @fredefox linked to, I'm wondering what the interaction might be between `collapseDefault` and infective options. `optTwoLevel` is `WithDefault 'False`. It's turned on when `cubicalFlag` is set:
https://github.com/agda/agda/blob/b51630a991e3b61e8aafd649f5b96ca23e80958b/src/full/Agda/Interaction/Options.hs#L763-L769

Though it does seem to be handled pretty similarly to `withoutK` (which is coinfective, not infective: was this difference intentional?)

**Comment by rwe** [comment]
Since this failure is in `master`, it affects all open PRs. Is there an easy way to temporarily suppress that specific error while this gets sorted out, so PRs like #4791, #4806, #4802, etc don't get caught in the crossfire of failed builds?

**Comment by Saizan** [comment]
Well, we could just turn off the `language-cubical` part of the tests about the manual, I think?

**Comment by rwe** [comment]
I've managed to reproduce this locally by deleting the `.agdai` files in `.stack-work` before running `make user-manual-test`:

```sh
find .stack-work -iname \*.agdai -print -delete
AGDA_TESTS_OPTIONS=-j2 BUILD_DIR="$(pwd)/$(stack path --dist-dir)" stack exec -- make user-manual-test
```

> ```
> agda» find .stack-work -iname \*.agdai -print -delete
> .stack-work/install/x86_64-osx/b5aef3850bf9908135bc66f7a5b82d928f7cb8cb62e5f6cbfe1c983ac87bd6a7/8.8.3/share/x86_64-osx-ghc-8.8.3/Agda-2.6.2/lib/prim/Agda/Primitive/Cubical.agdai
> .stack-work/install/x86_64-osx/b5aef3850bf9908135bc66f7a5b82d928f7cb8cb62e5f6cbfe1c983ac87bd6a7/8.8.3/share/x86_64-osx-ghc-8.8.3/Agda-2.6.2/lib/prim/Agda/Primitive.agdai
> .stack-work/install/x86_64-osx/b5aef3850bf9908135bc66f7a5b82d928f7cb8cb62e5f6cbfe1c983ac87bd6a7/8.8.3/share/x86_64-osx-ghc-8.8.3/Agda-2.6.2/lib/prim/Agda/Builtin/List.agdai
> .stack-work/install/x86_64-osx/b5aef3850bf9908135bc66f7a5b82d928f7cb8cb62e5f6cbfe1c983ac87bd6a7/8.8.3/share/x86_64-osx-ghc-8.8.3/Agda-2.6.2/lib/prim/Agda/Builtin/Nat.agdai
> .stack-work/install/x86_64-osx/b5aef3850bf9908135bc66f7a5b82d928f7cb8cb62e5f6cbfe1c983ac87bd6a7/8.8.3/share/x86_64-osx-ghc-8.8.3/Agda-2.6.2/lib/prim/Agda/Builtin/Bool.agdai
> .stack-work/install/x86_64-osx/b5aef3850bf9908135bc66f7a5b82d928f7cb8cb62e5f6cbfe1c983ac87bd6a7/8.8.3/share/x86_64-osx-ghc-8.8.3/Agda-2.6.2/lib/prim/Agda/Builtin/Coinduction.agdai
> .stack-work/install/x86_64-osx/b5aef3850bf9908135bc66f7a5b82d928f7cb8cb62e5f6cbfe1c983ac87bd6a7/8.8.3/share/x86_64-osx-ghc-8.8.3/Agda-2.6.2/lib/prim/Agda/Builtin/Sigma.agdai
> .stack-work/install/x86_64-osx/b5aef3850bf9908135bc66f7a5b82d928f7cb8cb62e5f6cbfe1c983ac87bd6a7/8.8.3/share/x86_64-osx-ghc-8.8.3/Agda-2.6.2/lib/prim/Agda/Builtin/Equality.agdai
> .stack-work/install/x86_64-osx/b5aef3850bf9908135bc66f7a5b82d928f7cb8cb62e5f6cbfe1c983ac87bd6a7/8.8.3/share/x86_64-osx-ghc-8.8.3/Agda-2.6.2/lib/prim/Agda/Builtin/Cubical/Path.agdai
> .stack-work/install/x86_64-osx/b5aef3850bf9908135bc66f7a5b82d928f7cb8cb62e5f6cbfe1c983ac87bd6a7/8.8.3/share/x86_64-osx-ghc-8.8.3/Agda-2.6.2/lib/prim/Agda/Builtin/Cubical/Sub.agdai
> agda» AGDA_TESTS_OPTIONS=-j2 BUILD_DIR="$(pwd)/$(stack path --dist-dir)" stack exec -- make user-manual-test
> ======================================================================
> ========================= User manual (test) =========================
> ======================================================================
> Using JS node binary at /usr/local/bin/node
> all
>   UserManual
>     language-cubical:                              FAIL (0.49s)
>                                                    Result did not match golden value.
>                                                    Diff between actual and golden value:
> ========================================
> agda-tests: Call to `git diff` failed: (ExitFailure 1,"diff --git a/private/var/folders/hz/_rph6srj7674fnwgcv48361w0000gn/T/language-cubical70182-0.golden b/private/var/folders/hz/_rph6srj7674fnwgcv48361w0000gn/T/language-cubical70182-1.actual\nindex 699465fef..28fb2ee81 100644\n--- a/private/var/folders/hz/_rph6srj7674fnwgcv48361w0000gn/T/language-cubical70182-0.golden\n+++ b/private/var/folders/hz/_rph6srj7674fnwgcv48361w0000gn/T/language-cubical70182-1.actual\n@@ -1,2 +1,10 @@\n-AGDA_SUCCESS\n+AGDA_UNEXPECTED_FAIL\n \n+ret > ExitFailure 42\n+out > agda-default-include-path/Agda/Builtin/Cubical/HCompU.agda:73,11-163\n+out > SSet\969 != SSet l\n+out > when checking that the inferred type of an application\n+out >   (i : I) \8594 e (~ i)\n+out > matches the expected type\n+out >   (i : I) \8594 e (~ i)\n+out >\n","")
> CallStack (from HasCallStack):
>   error, called at ./Test/Tasty/Silver/Interactive.hs:189:16 in tasty-silver-3.1.13-Izr5m8Ws97Q4ZDEapEISiY:Test.Tasty.Silver.Interactive
> make: *** [user-manual-test] Error 1
> ```

**Comment by rwe** [comment]
Here's a simplified script that will fail or succeed on demand:
```sh
agda_data_dir="$(stack exec -- ghc-pkg --simple-output field Agda data-dir)"
[ -n "${agda_data_dir}" ] || exit 1

clear_agdai() {
  find "${agda_data_dir}" -iname '*.agdai' -print -delete
}

compile_hcompu_no_flags() {
  stack exec -- agda --no-default-libraries "${agda_data_dir}/lib/prim/Agda/Builtin/Cubical/HCompU.agda"
}

run_tests() {
  AGDA_TESTS_OPTIONS='-j2' BUILD_DIR="$(pwd)/$(stack path --dist-dir)" stack exec -- make user-manual-test
}

echo 'The tests fail when clean'
clear_agdai && run_tests
echo "Status (expecting failure): $?"
echo 'And running again does not help.'
run_tests
echo "Status (expecting failure): $?"

echo 'But if HCompU is compiled first, they succeed:'
clear_agdai && compile_hcompu_no_flags && run_tests
echo "Status (expecting 0): $?"
echo 'Running again still succeeds:'
run_tests
echo "Status (expecting 0): $?"
```

You can put that into a script and run it.

It would be interesting to see the difference between the `.agdai` left over after `compile_hcompu_no_flags` and `run_tests` and see what the differences are, but I'm not sure how to inspect those.

**Comment by Saizan** [comment]
Thanks! That really helped, essentially `--double-check` is not happy with the code in `HCompU` as written.

It would be great if the testsuite printed out the actual agda command run.

**Comment by rwe** [comment]
Awesome! Yeah, I agree, I'm taking stabs in the dark trying to understand how things get run :)

**Comment by rwe** [comment]
And you're right! `--double-check` causes it to fail without any of the test stuff:
> ```
> find "${data_dir}" -iname '*.agdai' -delete && stack exec -- agda --no-default-libraries --double-check "${agda_data_dir}/lib/prim/Agda/Builtin/Cubical/HCompU.agda"
> 
> SSetω != SSet l
> when checking that the inferred type of an application
>   (i : I) → e (~ i)
> matches the expected type
>   (i : I) → e (~ i)
> ```

Glad that's narrowed down!

**Comment by Saizan** [comment]
If the tests pass on the `issue4805` branch then that should be good enough to merge, and I'll open another ticket for the underlying problem.

**Comment by rwe** [comment]
I tried manually adding the change from that branch (https://github.com/agda/agda/commit/b3b88ab4fd96d5c8ea32cbf55c45b8baebf9ac23) and ran the `stack exec…` command above and failed locally with:
```
Failed to solve the following constraints:
  outS b = outS b : e (~ i0) (blocked on _187)
  _u_185 (φ = i1) _ = transp (λ j → e (j ∧ ~ i0)) i0 (a _) : e (~ i0)
    (blocked on _u_185)
  transp e i0 (a _) = _u_185 (φ = i1) _ : e i1 (blocked on _u_185)
Unsolved metas at the following locations:
  …/lib/prim/Agda/Builtin/Cubical/HCompU.agda:78,139-143
  …/lib/prim/Agda/Builtin/Cubical/HCompU.agda:78,151-155
  …/lib/prim/Agda/Builtin/Cubical/HCompU.agda:78,156-157
  …/lib/prim/Agda/Builtin/Cubical/HCompU.agda:78,151-157
```

**Comment by Saizan** [comment]
Hopefully it works now, I'm working from a weird setup atm so I'm committing through github and missed some changes.

**Comment by rwe** [comment]
That appears to work! The whole `make user-manual-test` suite passes for me locally with your latest change.

**Comment by asr** [comment]
> I'll open another ticket for the underlying problem.

Did you open the ticket?

This issue showed two fails in our CI framework: We are not erasing all the necessary `*.agdai` files and we need to test if using `--double-check` is an option.

**Comment by Saizan** [comment]
Actually we already have #4455 about `--double-check` and `cubical`, which is what I meant by underlying problem.

I suppose this issue can be about the testsuite stuff then?

Btw, we should also consider the use of `--ignore-all-interfaces` if that's how we want the tests run.

And have the tests print out the agda flags when a test fail would be great.

**Comment by andreasabel** [comment]
Anything left to to here?  Otherwise, please close.

**Comment by asr** [comment]
> This issue showed two fails in our CI framework: We are not erasing all the necessary `*.agdai` files and we need to test if using `--double-check` is an option.

The issue related to erase the all *.agdai files should have been fixed by #5010. The issue for the  `--double-check` is #3099.
